### PR TITLE
improve errors for complex derivs

### LINKED ISSUE #3121: Error in backward-mode differentiation of ℝ → ℂ function
Hi,
First of all I'd like to thank you for the wonderful work you put forward with Jax. 
We recently started using it in [netket](https://github.com/netket/netket), a package for solving quantum-many body problems with neural-networks, and have been very pleasantly surprised by the performance and ease of use, so much that we rewrote most of our code to take advantage of it!

However, when dealing with wavefunctions we have to work a lot with complex functions. It is within this context that I noticed a silent error arises when trying to use reverse-mode over a ℝ → ℂ function (so, `grad` and `jacrev`).

Consider the following case:
```python
import jax
w=(jax.np.array([1.0]), jax.np.array([0.5]))

jax.grad(lambda w: w[0].sum() + 1j* w[1].sum(), holomorphic=True)(w)
(DeviceArray([1.], dtype=float32), DeviceArray([0.], dtype=float32))

 jax.jacrev(lambda w: w[0].sum() + 1j* w[1].sum(), holomorphic=True)(w)
(DeviceArray([1.], dtype=float32), DeviceArray([0.], dtype=float32))
```
- The gradient is wrong, it should be complex-valued. I have the feeling you are probably competing the right gradient but you cast it at the end to the original type, resulting in this error.

As a temporary fix, I figured out I can just promote types to be complex and compute the gradient by saying that the function is holomorphic, and re-scaling the complex part, though it would be nice if this could be fixed.
```python 
wc=(jax.np.array([1.0], dtype=jax.np.complex64), jax.np.array([0.5], dtype=jax.np.complex64))

jax.jacrev(lambda w: w[0].sum() + 1j* w[1].sum(), holomorphic=True)(wc)
(DeviceArray([1.-0.j], dtype=complex64), DeviceArray([0.+1.j], dtype=complex64))
```

Slightly related,  I understand why you ask the user to mark the function to be holomorphic once you hit a complex result, but in this case it's a bit weird: ℝ → ℂ functions are always holomorphic. I don't know if anything can be done about it.

cc @gcarleo

### PR DESCRIPTION
**Pull request by mattjj**
fixes #3121

### CODE REVIEW

**Comment by mattjj** [comment]
Since this is just an improvement on error messages, I'm going to merge now, but will iterate on any suggested revisions/improvements in subsequent PRs!
